src/shm/shm.c

```diff
@@ -0,0 +1,470 @@
+    printf("thr-c %p\t%p\t%p\t%p\n", arg->filename, arg->coord, arg->copy, arg->writer_ready);
+
+    cond_package* wr_ready = arg->writer_ready;
+    copy_struct* copy_mem = arg->copy;
+
+    // To do before allowing readers to join:
+    // + Take writing lock on the first active segment (will allow them to queue in waiting[1])
+    // + Take lock on reader_cnt (will prevent them from updating it)
+    // + Set started flag to false
+    int active_segment = 0;
+    pthread_rwlock_wrlock(&(copy_mem->active[active_segment]));
+    printf("\tWrite lock on active %d\n", active_segment);
+    pthread_mutex_lock(&(copy_mem->reader_cnt.lock));
+    
+    pthread_mutex_lock(&(copy_mem->started_lock));
+    copy_mem->started_flag = 0;
+    pthread_mutex_unlock(&(copy_mem->started_lock));
+
+    // Now wake up the parent thread, readers will start joining, queued by joining_lock and the
+    // first will wait on reader_cnt
+    pthread_mutex_lock(&(wr_ready->lock));
+    wr_ready->v = 1;
+    pthread_cond_signal(&(wr_ready->cond)); // Signals the parent threat to start letting 
+                                            // readers in, they'll be blocked by the sync_var lock
+    
+    int r = pthread_mutex_unlock(&(wr_ready->lock)); // Allow parent to wake up
+    printf("unlckw %d %d\n", r, errno);
+
+    // ACTIVE LOCKS: active[0], reader_cnt
+    copy_mem->reader_cnt.v = 0;
+    // Repeat until enough joined:
+    // 1. Signal to signal_rdr that it can continue (first iteration this will be pointless)
+    // 2. Wait on reader_cnt (unlocking it)
+    // If it exited before signaling reader_cnt I need to:
+    // + Update the begin flag
+    // + Signal the reader
+    while (copy_mem->reader_cnt.v < NUM_READERS){
+        printf("\tWaiting for readers\n");
+        pthread_mutex_lock(&(copy_mem->signal_rdr.lock));
+            copy_mem->signal_rdr.v = 1;
+            pthread_cond_signal(&(copy_mem->signal_rdr.cond));
+        pthread_mutex_unlock(&(copy_mem->signal_rdr.lock)); // Will let the next reader join but it'll block on cnt
+
+        pthread_cond_wait(&(copy_mem->reader_cnt.cond), &(copy_mem->reader_cnt.lock));
+        printf("\tHave %d readers\n", copy_mem->reader_cnt.v);
+    }
+    // Here we can start
+    pthread_mutex_lock(&(copy_mem->started_lock));
+        copy_mem->started_flag = 1;
+    pthread_mutex_unlock(&(copy_mem->started_lock));
+
+    pthread_mutex_lock(&(copy_mem->signal_rdr.lock));
+        copy_mem->signal_rdr.v = 1;
+        pthread_cond_signal(&(copy_mem->signal_rdr.cond));
+    pthread_mutex_unlock(&(copy_mem->signal_rdr.lock)); // Will let the next reader join but it'll fail the begin check
+    
+    // ACTIVE LOCKS: active[0], reader_cnt
+    
+    // Before starting wake up the parent thread (pretending to be a new reader)
+    coord_struct* coord = arg->coord;
+    pthread_mutex_lock(&(coord->reader_arrived.lock));
+    coord->reader_arrived.v = 1;
+    pthread_cond_signal(&(coord->reader_arrived.cond));
```

Comment:
I indented some blocks to signal which areas where operating protected by a mutex, however I should fix it because I was anyway inconsistent in the application of this style and comments are clearer

In the reworked simplifies shared memory implementation I tend to write comments explicitly stating which locks are held at some point, I should do the same here